### PR TITLE
Fix PR #963 : Passing all the params through the apply call in lda.get_document_topics

### PR DESCRIPTION
**Pull request by parthoiiitm**
Hi, I have added the test case to check get_document_topics() working on corpus or not. However, the build will fail because of the new test case I have added.  For some words in corpus, `word_topics` and `word_phis` are not returned .

If the corpus is small, I am getting the correct value. However, if I am taking the corpus as defined in `test_ldamodel.py`, then it throws the error  . Let me show this via a toy example called `app.py`

```
from gensim.corpora import mmcorpus, Dictionary
from gensim import models


texts = [['human', 'interface', 'computer'],
['survey', 'user', 'computer', 'system', 'response', 'time'],
['eps', 'user', 'interface', 'system']]
dictionary = Dictionary(texts)
corpus = [dictionary.doc2bow(text) for text in texts]

lda = models.LdaModel(corpus)

# ab = lda.get_document_topics(corpus)
# for t in ab:
#     print('new doc')
#     print(t)
#
# print len(ab)
# print len(corpus)

doc_topics, word_topics, word_phis = lda.get_document_topics(corpus, per_word_topics=True, minimum_probability=0.02)

for t in doc_topics:
    print('new doc_topic')
    print(t)

for t in word_topics:
    print('new word_topic')
    print(t)

for t in word_phis:
    print('new word_phis')
    print(t)
```

For the above code, I am getting complete output as follows:

```
new doc_topic
[(96, 0.75249999999999984)]
new doc_topic
[(0, [96]), (1, [96]), (2, [96])]
new doc_topic
[(0, [(96, 0.99999999999999989)]), (1, [(96, 1.0)]), (2, [(96, 1.0)])]
new word_topic
[(2, 0.85857142857143109)]
new word_topic
[(1, [2]), (3, [2]), (4, [2]), (5, [2]), (6, [2]), (7, [2])]
new word_topic
[(1, [(2, 1.0)]), (3, [(2, 1.0)]), (4, [(2, 1.0)]), (5, [(2, 1.0)]), (6, [(2, 1.0)]), (7, [(2, 1.0)])]
new word_phis
[(29, 0.80200000000000238)]
new word_phis
[(0, [29]), (6, [29]), (7, [29]), (8, [29])]
new word_phis
[(0, [(29, 0.99999999999999989)]), (6, [(29, 1.0)]), (7, [(29, 0.99999999999999989)]), (8, [(29, 0.99999999999999989)])]
```

However, if I take the corpus as:

```
texts = [['human', 'interface', 'computer'],
['survey', 'user', 'computer', 'system', 'response', 'time'],
['eps', 'user', 'interface', 'system'],
['system', 'human', 'system', 'eps'],
['user', 'response', 'time'],
['trees'],
['graph', 'trees'],
['graph', 'minors', 'trees'],
['graph', 'minors', 'survey']]
```

then the following error will be thrown:

```
/Users/partho/anaconda/bin/python /Users/partho/PycharmProjects/untitled/app.py
Traceback (most recent call last):
  File "/Users/partho/PycharmProjects/untitled/app.py", line 39, in <module>
    doc_topics, word_topics, word_phis = lda.get_document_topics(corpus, per_word_topics=True, minimum_probability=0.02)
ValueError: too many values to unpack
```

Please let me know how to modify the code for word_topics and word_phis in the returned tuple so that it won't throw the ValueError.

### CODE REVIEW

**Comment by tmylk** [comment]
is it just that they are [filtered](https://github.com/parthoiiitm/gensim/blob/102393cf0add37e312f7f58c933afe5811ef7d98/gensim/models/ldamodel.py#L944)?

Feel free to change the `minimum_phi_value` passed

**Comment by parthoiiitm** [comment]
Hi @tmylk ,
I tried to change ```minimum_phi_value``` , but it didn't made much difference. I think I got where the problem is coming. The problem is coming at corpus index  3, that is
 ```['system', 'human', 'system', 'eps']``` . 

I guess there's mismatch in the actual corpus mentioned and the internal corpus generated, i.e., 

```
corpus document at index 3 : ['system', 'human', 'system', 'eps']
internal corpus document at index 3: [(2, 1), (6, 2), (8, 1)]
``` 
I printed the generated  doc_topics, word_topics and word_phis values before the tuple is returned. For test purpose, I changed num_topics to 3.  

```
For document in corpus : [(0, 1), (1, 1), (2, 1)]
  
Document Topics:  [(0, 0.085256386514654198), (1, 0.82642252962016716), (2, 0.088321083865178618)]
Word Topics:  [(0, [1, 2, 0]), (1, [1, 2, 0]), (2, [1, 2, 0])]
Word Phis:  [(0, [(0, 0.0025285850037330788), (1, 0.99076304456354658), (2, 0.0067083704327202329)]), (1, [(0, 0.0025404420644047185), (1, 0.99068360988050452), (2, 0.0067759480550907752)]), (2, [(0, 0.0026214330077555916), (1, 0.99099459056294825), (2, 0.0063839764292960774)])]
  
------------------------------------------------
  
For document in corpus : [(1, 1), (3, 1), (4, 1), (5, 1), (6, 1), (7, 1)]
  
Document Topics:  [(0, 0.049812014409508887), (1, 0.052621163472570683), (2, 0.89756682211792049)]
Word Topics:  [(1, [2, 1, 0]), (3, [2, 0, 1]), (4, [2, 0, 1]), (5, [2, 0, 1]), (6, [2, 0, 1]), (7, [2, 0, 1])]
Word Phis:  [(1, [(0, 0.0035630465521965778), (1, 0.027604390510358127), (2, 0.96883256293744524)]), (3, [(0, 0.0030831897263170261), (1, 0.001872865014820153), (2, 0.99504394525886286)]), (4, [(0, 0.0030848566560638716), (1, 0.0019008301229193109), (2, 0.99501431322101674)]), (5, [(0, 0.0029858846340732647), (1, 0.001775116048036694), (2, 0.99523899931788995)]), (6, [(0, 0.001129649006438651), (1, 0.00066075451665730445), (2, 0.99820959647690399)]), (7, [(0, 0.0014994326582106006), (1, 0.00092524514381757015), (2, 0.99757532219797196)])]
  
------------------------------------------------
  
For document in corpus : [(0, 1), (6, 1), (7, 1), (8, 1)]
  
Document Topics:  [(0, 0.068962110663396636), (1, 0.077321753048242792), (2, 0.85371613628836052)]
Word Topics:  [(0, [2, 1, 0]), (6, [2, 0, 1]), (7, [2, 0, 1]), (8, [2, 0, 1])]
Word Phis:  [(0, [(0, 0.005169442922469107), (1, 0.048359110186989941), (2, 0.94647144689054097)]), (6, [(0, 0.0016669760059777764), (1, 0.0011717562993173935), (2, 0.9971612676947047)]), (7, [(0, 0.0022118074173394261), (1, 0.0016401685918843076), (2, 0.99614802399077607)]), (8, [(0, 0.0024277508338771689), (1, 0.0017138915575534686), (2, 0.99585835760856922)])]
  
------------------------------------------------
  
For document in corpus : [(2, 1), (6, 2), (8, 1)]
  
Document Topics:  [(0, 0.068940203341492784), (1, 0.077973801355772726), (2, 0.85308599530273455)]
Word Topics:  [(2, [2, 1, 0]), (6, [2, 0, 1]), (8, [2, 0, 1])]
Word Phis:  [(2, [(0, 0.0056065227731108523), (1, 0.051937673728930898), (2, 0.94245580349795821)]), (6, [(0, 0.0033331555062166588), (1, 0.0024047924039207694), (2, 1.9942620520898626)]), (8, [(0, 0.0024271367641137843), (1, 0.0017586829106569111), (2, 0.99581418032522939)])]
  
Traceback (most recent call last):
  File "/Users/partho/PycharmProjects/untitled/app.py", line 39, in <module>
    doc_topics, word_topics, word_phis = lda.get_document_topics(corpus, per_word_topics=True, minimum_probability=0.02, minimum_phi_value=0.0)
ValueError: too many values to unpack

------------------------------------------------
  ```

Can you please guide me why this mismatch is happening internally ??

**Comment by tmylk** [comment]
So what is happening in topic 3?

**Comment by parthoiiitm** [comment]
At [this line](https://github.com/parthoiiitm/gensim/blob/102393cf0add37e312f7f58c933afe5811ef7d98/gensim/models/ldamodel.py#L944), we are returning ```(document_topics, word_topic, word_phi)``` as a tuple. So, if you call the following function:
```doc_topics, word_topic, word_phis = lda.get_document_topics(corpus, per_word_topics=True)```, you are basically storing each tuple in a particular variable, i.e., variable ``` doc_topics```  store tuple data ```(document_topics, word_topic, word_phi)``` for document 1, and variables ```word_topic``` and ```word_phis``` store tuple data for doucment 2 and 3. Since,
```lda.get_document_topics(corpus, per_word_topics=True)``` returns list of tuples and the number of these tuples is more than 3, and 3 variables aren't enough to store these data in case your corpus is greater than 3 😃 

I will show below for others to realize what mistake they are making while making call to function.

```
from gensim.corpora import mmcorpus, Dictionary
from gensim import models


texts = [['human', 'interface', 'computer'],
['survey', 'user', 'computer', 'system', 'response', 'time'],
['eps', 'user', 'interface', 'system'],
['system', 'human', 'system','eps'],
['user', 'response', 'time'],
['trees'],
['graph', 'trees'],
['graph', 'minors', 'trees'],
['graph', 'minors', 'survey']]

dictionary = Dictionary(texts)
corpus = [dictionary.doc2bow(text) for text in texts]

lda = models.LdaModel(corpus)

topics, a, b = lda.get_document_topics(corpus, per_word_topics=True)

print (document_topics)
print ( word_topic)
print ( word_phi)

```
will output :
```
[[(19, 0.75250000000000328)], [(0, [19]), (1, [19]), (2, [19])], [(0, [(19, 1.0)]), (1, [(19, 1.0)]), (2, [(19, 0.99999999999999989)])]]

[[(86, 0.85857142576872991)], [(1, [86]), (3, [86]), (4, [86]), (5, [86]), (6, [86]), (7, [86])], [(1, [(86, 1.0)]), (3, [(86, 1.0)]), (4, [(86, 1.0)]), (5, [(86, 1.0)]), (6, [(86, 1.0)]), (7, [(86, 1.0)])]]

[[(77, 0.80200000000000071)], [(0, [77]), (6, [77]), (7, [77]), (8, [77])], [(0, [(77, 1.0)]), (6, [(77, 1.0)]), (7, [(77, 1.0)]), (8, [(77, 1.0)])]]
```
As we can see, each tuple has 3 lists of ```document_topics```, ```word_topic```, ```word_phi```.

So, the proper way to access the data for the entire corpus is like this :

``` 
topics = lda.get_document_topics(corpus, per_word_topics=True)
for t in topics:
    print('new doc')
    print 'Document topics:', t[0]
    print 'Word topics:', t[1]
    print 'Word phis:', t[2]
    print('--------------')
```

which will return:

```
new doc
Document topics: [(80, 0.75250000000000072)]
Word topics: [(0, [80]), (1, [80]), (2, [80])]
Word phis: [(0, [(80, 0.99999999999999989)]), (1, [(80, 0.99999999999999989)]), (2, [(80, 1.0)])]
--------------
new doc
Document topics: [(56, 0.85857142857142965)]
Word topics: [(1, [56]), (3, [56]), (4, [56]), (5, [56]), (6, [56]), (7, [56])]
Word phis: [(1, [(56, 1.0)]), (3, [(56, 1.0)]), (4, [(56, 1.0)]), (5, [(56, 1.0)]), (6, [(56, 1.0)]), (7, [(56, 1.0)])]
--------------
new doc
Document topics: [(93, 0.80200000000000016)]
Word topics: [(0, [93]), (6, [93]), (7, [93]), (8, [93])]
Word phis: [(0, [(93, 1.0)]), (6, [(93, 1.0)]), (7, [(93, 1.0)]), (8, [(93, 1.0)])]
--------------
new doc
Document topics: [(80, 0.80200000000000049)]
Word topics: [(2, [80]), (6, [80]), (8, [80])]
Word phis: [(2, [(80, 1.0)]), (6, [(80, 2.0)]), (8, [(80, 0.99999999999999989)])]
--------------
new doc
Document topics: [(1, 0.75250000000000394)]
Word topics: [(3, [1]), (4, [1]), (7, [1])]
Word phis: [(3, [(1, 1.0)]), (4, [(1, 0.99999999999999989)]), (7, [(1, 1.0)])]
--------------
                                              .
                                              .
                                              .
```

**Comment by parthoiiitm** [comment]
In order to get the `doc_topics`, `word_topics` and `word_phis` of the overall corpus, I restructured the output data by creating a helper function. This function is very inefficient in terms of complexity as of now, but for testing purpose, I want to confirm whether the output of the aggregated result is desirable or not. If yes, then I will work on making this helper function more efficient so that we can include in the final code.

The helper function is as follows:

```
def get_document_topics_for_corpus(topics):
    document_topics = dict()
    word_topics = dict()
    word_phis = dict()

    doc_topics = list()
    word_top = list()
    word_ph = list()

    for doc_topic, word_topic, word_phi in topics:
        #Document_topics aggregation
        key_doc = doc_topic[0][0]
        value_doc = doc_topic[0][1]
        document_topics.setdefault(key_doc, value_doc)

        #Word_topics aggregation
        for key in  word_topic:
            word_topics.setdefault(key[0], [])
            word_topics[key[0]].append(key_doc)

        #Word_phis aggregation
        for key in word_phi:
            word_phis.setdefault(key[0], [])
            word_phis[key[0]].append(key[1][0])

    for key, value in document_topics.iteritems():
        temp = (key, value)
        doc_topics.append(temp)

    for key, value in word_topics.iteritems():
        temp = (key, value)
        word_top.append(temp)

    for key, value in word_phis.iteritems():
        temp = (key, value)
        word_ph.append(temp)


    return (doc_topics, word_top, word_ph)
```

And now,  I can call the above function as follows:
```
topics = lda.get_document_topics(corpus, per_word_topics=True)
doc_topics, word_topics, word_phis = get_document_topics_for_corpus(topics)

print "Document topics: ", doc_topics
print "Word topics: ", word_topics
print "Word phis:", word_phis

```

The actual result is as follows:
```
new doc
Document topics: [(79, 0.75250000000000072)]
Word topics: [(0, [79]), (1, [79]), (2, [79])]
Word phis: [(0, [(79, 1.0)]), (1, [(79, 1.0)]), (2, [(79, 1.0)])]
--------------
new doc
Document topics: [(23, 0.85857142857143054)]
Word topics: [(1, [23]), (3, [23]), (4, [23]), (5, [23]), (6, [23]), (7, [23])]
Word phis: [(1, [(23, 1.0)]), (3, [(23, 1.0)]), (4, [(23, 1.0)]), (5, [(23, 1.0)]), (6, [(23, 1.0)]), (7, [(23, 1.0)])]
--------------
new doc
Document topics: [(28, 0.80199999993851401)]
Word topics: [(0, [28]), (6, [28]), (7, [28]), (8, [28])]
Word phis: [(0, [(28, 1.0)]), (6, [(28, 1.0)]), (7, [(28, 1.0000000000000002)]), (8, [(28, 1.0)])]
--------------
new doc
Document topics: [(37, 0.80200000000000204)]
Word topics: [(2, [37]), (6, [37]), (8, [37])]
Word phis: [(2, [(37, 1.0)]), (6, [(37, 2.0)]), (8, [(37, 1.0)])]
--------------
new doc
Document topics: [(23, 0.7524999986233426)]
Word topics: [(3, [23]), (4, [23]), (7, [23])]
Word phis: [(3, [(23, 1.0)]), (4, [(23, 1.0)]), (7, [(23, 1.0)])]
--------------
new doc
Document topics: [(74, 0.50499999999999856)]
Word topics: [(9, [74])]
Word phis: [(9, [(74, 1.0)])]
--------------
new doc
Document topics: [(40, 0.6700000000000027)]
Word topics: [(9, [40]), (10, [40])]
Word phis: [(9, [(40, 1.0)]), (10, [(40, 1.0)])]
--------------
new doc
Document topics: [(42, 0.75250000000000228)]
Word topics: [(9, [42]), (10, [42]), (11, [42])]
Word phis: [(9, [(42, 1.0)]), (10, [(42, 1.0)]), (11, [(42, 1.0)])]
--------------
new doc
Document topics: [(62, 0.7525000000000015)]
Word topics: [(5, [62]), (10, [62]), (11, [62])]
Word phis: [(5, [(62, 1.0)]), (10, [(62, 1.0)]), (11, [(62, 1.0)])]
--------------
```
**And the aggregated result are as follows:**

```
Document topics:  [(37, 0.80200000000000204), (40, 0.6700000000000027), (74, 0.50499999999999856), (79, 0.75250000000000072), (23, 0.85857142857143054), (28, 0.80199999993851401), (42, 0.75250000000000228), (62, 0.7525000000000015)]

Word topics:  [(0, [79, 28]), (1, [79, 23]), (2, [79, 37]), (3, [23, 23]), (4, [23, 23]), (5, [23, 62]), (6, [23, 28, 37]), (7, [23, 28, 23]), (8, [28, 37]), (9, [74, 40, 42]), (10, [40, 42, 62]), (11, [42, 62])]

Word phis: [(0, [(79, 1.0), (28, 1.0)]), (1, [(79, 1.0), (23, 1.0)]), (2, [(79, 1.0), (37, 1.0)]), (3, [(23, 1.0), (23, 1.0)]), (4, [(23, 1.0), (23, 1.0)]), (5, [(23, 1.0), (62, 1.0)]), (6, [(23, 1.0), (28, 1.0), (37, 2.0)]), (7, [(23, 1.0), (28, 1.0000000000000002), (23, 1.0)]), (8, [(28, 1.0), (37, 1.0)]), (9, [(74, 1.0), (40, 1.0), (42, 1.0)]), (10, [(40, 1.0), (42, 1.0), (62, 1.0)]), (11, [(42, 1.0), (62, 1.0)])]

```

Is the aggregated result desirable? If yes, then I will work on improving this helper function. Can you also give me some tips in improving this helper function ??

**Comment by tmylk** [comment]
I don't see a need for aggregated function - please convince me why you think it is heplful.

The current API below looks ok, though it needs to be documented in the docstring and in the lda ipynb tutorial
```
all_topics = lda.get_document_topics(corpus, per_word_topics=True)
for doc_topic, word_topic, word_phi in all_topics:
```

**Comment by parthoiiitm** [comment]
@tmylk Yes, I do agree that the aggregated function is not needed actually, but users were confused and calling like this:
```
doc_topics, word_topics, word_phis  = lda.get_document_topics(corpus, per_word_topics=True)
```
instead of this:

```
topics =  lda.get_document_topics(corpus, per_word_topics=True)
```
It will be a good idea to be documented in docstring and lda tutorial.

Plus, users can store the data for all corpus like this:
```
topics =  lda.get_document_topics(corpus, per_word_topics=True)
all_topics = [(doc_topics, word_topics, word_phis) for doc_topics, word_topics, word_phis in topics]
```
And access a particular document details like this:
```
print all_topics[doc_index]
```

So, will this PR gets merged with the develop branch? I haven't included the aggregated function in the coce yet.

**Comment by tmylk** [comment]
Sounds great. Let's update docstring and tutorial.

**Comment by parthoiiitm** [comment]
Can you let me know which files to update in docstring and tutorial? I will commit the changes, and you can review the code supposed to be merged.

File: gensim/models/ldamodel.py [hunk 1]
```diff
@@ -977,7 +985,9 @@ def __getitem__(self, bow, eps=None):
         Ignore topics with very low probability (below `eps`).
 
         """
-        return self.get_document_topics(bow, eps)
+        #Is eps equivalent to minimum_probability?
+        eps = self.minimum_probability
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason to add this line? it is used anyway in https://github.com/parthoiiitm/gensim/blob/2ac2f951452e7d968acace2b6d7b11be8b79ad62/gensim/models/ldamodel.py#L912

File: gensim/models/ldamulticore.py [hunk 2]
```diff
@@ -80,7 +80,8 @@ class LdaMulticore(LdaModel):
     def __init__(self, corpus=None, num_topics=100, id2word=None, workers=None,
                  chunksize=2000, passes=1, batch=False, alpha='symmetric',
                  eta=None, decay=0.5, offset=1.0, eval_every=10, iterations=50,
-                 gamma_threshold=0.001, random_state=None):
+                 gamma_threshold=0.001, random_state=None,
```
Comment (tmylk) [inline gensim/models/ldamulticore.py]: where is `minimium_probability`? Please add it

File: gensim/test/test_ldamodel.py [hunk 3]
```diff
@@ -249,6 +249,26 @@ def testGetDocumentTopics(self):
                 self.assertTrue(isinstance(k, int))
                 self.assertTrue(isinstance(v, float))
 
+        #Test case to use the get_document_topic function for the corpus
```
Comment (tmylk) [inline gensim/test/test_ldamodel.py]: Please add a new test for `minimum_prob` and `minimum_phi_value`. Select values that change the number of topics and docs returned

File: gensim/test/test_ldamodel.py [hunk 4]
```diff
@@ -249,6 +249,26 @@ def testGetDocumentTopics(self):
...
+        all_topics = model.get_document_topics(self.corpus, minimum_probability=0.5, minimum_phi_value=0.3, per_word_topics=True)
+        
+        self.assertEqual(model.state.numdocs, len(corpus))
```
Comment (tmylk) [inline gensim/test/test_ldamodel.py]: so the filtering params have no effect?

**Pull request merged**